renku/ui/api/models/activity.py

```diff
@@ -0,0 +1,427 @@
+
+    @property
+    def following_activities(self) -> List["Activity"]:
+        """Return a list of downstream activities."""
+        activity_gateway: Optional[ActivityGateway] = get_activity_gateway()
+        if activity_gateway is None:
+            return []
+
+        with replace_injection({IDatabaseDispatcher: activity_gateway.database_dispatcher}):
+            return [Activity.from_activity(a) for a in activity_gateway.get_downstream_activities(self._activity)]
+
+    @property
+    def parameters(self) -> List["FieldValue"]:
+        """Return list of parameters in this activity."""
+        if self._parameters is None:
+            self._parameters = [
+                FieldValue.from_parameter_value(p, self._activity.association.plan) for p in self._activity.parameters
+            ]
+
+        return self._parameters
+
+    @property
+    def plan(self) -> Plan:
+        """Activity's base plan."""
+        if self._plan is None:
+            self._plan = Plan.from_plan(self._activity.association.plan)
+
+        return self._plan
+
+
+class UsedInput(NamedTuple):
+    """Represent an input used in an execution."""
+
+    path: str
+    checksum: str
+
+    @classmethod
+    def from_usage(cls, usage: core_activity.Usage):
+        """Create an instance from core Usage."""
+        return cls(path=usage.entity.path, checksum=usage.entity.checksum)
+
+    def __repr__(self):
+        return f"<Input '{self.path}'>"
+
+
+class CreatedOutput(NamedTuple):
+    """Represent an output generated in an execution."""
+
+    path: str
+    checksum: str
+
+    @classmethod
+    def from_generation(cls, generation: core_activity.Generation):
+        """Create an instance from core Generation."""
+        return cls(path=generation.entity.path, checksum=generation.entity.checksum)
+
+    def __repr__(self):
+        return f"<Output '{self.path}'>"
+
+
+class FieldValue(NamedTuple):
+    """Represent a parameter used in an execution."""
+
+    field: Union[Input, Output, Parameter]
+    value: Any
+
+    @classmethod
+    def from_parameter_value(cls, parameter: core_parameter.ParameterValue, plan: core_plan.Plan):
+        """Create an instance from core ParameterValue."""
+        return cls(field=convert_parameter(plan.get_field_by_id(parameter.parameter_id)), value=parameter.value)
+
+    def __repr__(self):
+        if isinstance(self.field, Input):
+            return f"<Input '{self.value}'>"
+        elif isinstance(self.field, Output):
+            return f"<Output '{self.value}'>"
+
+        return f"<Parameter '{self.field.name}'={self.value}>"
+
+
+def get_activities(plan_id: str = None) -> List[Activity]:
+    """Return list of activities from a given ``plan_id``."""
+    activity_gateway = get_activity_gateway()
+    if not activity_gateway:
+        return []
+
+    activities = activity_gateway.get_all_activities()
+
+    if plan_id:
+        activities = [a for a in activities if a.association.plan.id == plan_id]
```

Comment:
Plans can be derived from, in which case their id changes. But we only allow (basically) non-functional changes. so I wonder if we should check for all instances of a plan here.

Reply:
I've change this to return all activities that use a plan or any of its upstream plans. To have less overhead, it won't include downstream plans. Users can still get the latest version of a plan and then get its activities to have all version. I can change it to get the latest version automatically if you think that's a better approach.